### PR TITLE
Wolfgang s feature/es api 5.0 v2

### PR DESCRIPTION
**Pull request by aquid**

### CODE REVIEW

**Comment by wolfgang-s** [comment]
@pulkitsinghal @aquid 
This looks promising, I think you can merge it.

Thanks!

File: test/02.basic-querying.test.js [hunk 1]
```diff
@@ -4,7 +4,7 @@ var db, User, Customer, AccessToken, Post, PostWithId, Category, SubCategory;
 
 /*eslint no-console: "off"*/
 /*global getSchema should*/
-describe('basic-querying', function () {
+xdescribe('basic-querying', function () {
```
Comment (pulkitsinghal) [inline test/02.basic-querying.test.js]: @aquid - i need to fix the tests still, correct?

Comment (aquid) [inline test/02.basic-querying.test.js]: yes @pulkitsinghal

**Comment by pulkitsinghal** [comment]
After extensively working on the failing tests yesterday, my guess is that:
1. either we have introduced a genuine bug
2. or we have exposed a race condition where the request order is:
  1. create index, delete index, if index exists true then put mapping
  2. but between `if index exists true` and `then put mapping` the index gets deleted

I am not comfortable releasing a versio nwhere the test are clearly failing. And honestly I don't necessarily have the time to get this out as quick as everyone really needs 5.x so folks will just have to sit tight, contribute to fixing the test or keep using their own forked branches :(

**Comment by pulkitsinghal** [comment]
@wolfgang-s I am still working on this, finding it very difficult to give it time but haven't given up

**Comment by wolfgang-s** [comment]
I can take a look at it next week.

No worries.

**Comment by pulkitsinghal** [comment]
@aquid and @wolfgang-s 

* I have added docs and folder structure to test all versions together or one-by-one.
* There is one unresolved test failure worth looking into for v2.x
```
TRACE: 2017-04-22T06:18:28Z
  -> POST http://localhost:9203/shakespeare/User/_update_by_query?refresh=true
  {
    "query": {
      "bool": {
        "must": [
          {
            "match": {
              "_id": 10
            }
          }
        ]
      }
    },
    "script": {
      "inline": "ctx._source.order= order;",
      "params": {
        "order": 10
      }
    }
  }
  <- 500
  {
    "error": {
      "root_cause": [
        {
          "type": "script_exception",
          "reason": "compile error",
          "script_stack": [
            "ctx._source.order= order;",
            "                   ^---- HERE"
          ],
          "script": "ctx._source.order= order;",
          "lang": "painless"
        }
      ],
      "type": "script_exception",
      "reason": "compile error",
      "script_stack": [
        "ctx._source.order= order;",
        "                   ^---- HERE"
      ],
      "script": "ctx._source.order= order;",
      "lang": "painless",
      "caused_by": {
        "type": "illegal_argument_exception",
        "reason": "Variable [order] is not defined."
      }
    },
    "status": 500
  }
```
* There are two unresolved test failures worth looking into for v5.x
```
TRACE: 2017-04-22T06:21:22Z
  -> POST http://localhost:9203/shakespeare/PostWithId/_search?size=50
  {
    "sort": [
      "id"
    ],
    "query": {
      "bool": {
        "must": [
          {
            "match": {
              "title": "Posts"
            }
          }
        ]
      }
    }
  }
  <- 400
  {
    "error": {
      "root_cause": [
        {
          "type": "illegal_argument_exception",
          "reason": "Fielddata is disabled on text fields by default. Set fielddata=true on [id] in order to load fielddata in memory by uninverting the inverted index. Note that this can however use significant memory."
        }
      ],
      "type": "search_phase_execution_exception",
      "reason": "all shards failed",
      "phase": "query",
      "grouped": true,
      "failed_shards": [
        {
          "shard": 0,
          "index": "shakespeare",
          "node": "BKqFuny1TleiQvY4kRfLug",
          "reason": {
            "type": "illegal_argument_exception",
            "reason": "Fielddata is disabled on text fields by default. Set fielddata=true on [id] in order to load fielddata in memory by uninverting the inverted index. Note that this can however use significant memory."
          }
        }
      ],
      "caused_by": {
        "type": "illegal_argument_exception",
        "reason": "Fielddata is disabled on text fields by default. Set fielddata=true on [id] in order to load fielddata in memory by uninverting the inverted index. Note that this can however use significant memory."
      }
    },
    "status": 400
  }
```

and 
```
TRACE: 2017-04-22T06:21:53Z
  -> POST http://localhost:9203/shakespeare/User/_update_by_query?refresh=true
  {
    "query": {
      "bool": {
        "must": [
          {
            "match": {
              "_id": 10
            }
          }
        ]
      }
    },
    "script": {
      "inline": "ctx._source.order= order;",
      "params": {
        "order": 10
      }
    }
  }
  <- 500
  {
    "error": {
      "root_cause": [
        {
          "type": "script_exception",
          "reason": "compile error",
          "script_stack": [
            "ctx._source.order= order;",
            "                   ^---- HERE"
          ],
          "script": "ctx._source.order= order;",
          "lang": "painless"
        }
      ],
      "type": "script_exception",
      "reason": "compile error",
      "script_stack": [
        "ctx._source.order= order;",
        "                   ^---- HERE"
      ],
      "script": "ctx._source.order= order;",
      "lang": "painless",
      "caused_by": {
        "type": "illegal_argument_exception",
        "reason": "Variable [order] is not defined."
      }
    },
    "status": 500
  }
```

**Comment by pulkitsinghal** [comment]
so @aquid and/or @wolfgang-s ... is either one of you upto fixing some of all of the 3 remaining errors? They may be simple mis-configurations but even fixing that would go a long way in making sure our tests are reliable.

**Comment by wolfgang-s** [comment]
Ok, I fixed the remaining errors. Next time I will start earlier ;)

- 1 error was cause by setting {id: true} on PostWithId (<- don't really now why this was there in the first place).

- script compilation errors for es5 are because it uses painless scripting language by default and not groovy (ES 1/2), so I changed this as well to use the variable "params".

patch file:

```
diff --git a/lib/esConnector.js b/lib/esConnector.js
index 30dea5a..8c39a72 100644
--- a/lib/esConnector.js
+++ b/lib/esConnector.js
@@ -948,7 +948,13 @@ ESConnector.prototype.updateAll = function updateAll(model, where, data, options
 	};
 	_.forEach(data, function (value, key) {
 		if (key !== '_id' || key !== idName) {
-			body.script.inline += 'ctx._source.' + key + '= ' + key + ';';
+			if(self.apiVersion[0] > 2) {
+				// default language for inline scripts is painless if ES 5, so this needs the extra params.
+                body.script.inline += 'ctx._source.' + key + '=params.' + key + ';';
+			} else {
+				// groovy syntax
+                body.script.inline += 'ctx._source.' + key + '=' + key + ';';
+			}
 			body.script.params[key] = value;
 		}
 	});
diff --git a/test/es-v5/03.basic-querying-no-timeout.test.js b/test/es-v5/03.basic-querying-no-timeout.test.js
index f6b6589..b03b362 100644
--- a/test/es-v5/03.basic-querying-no-timeout.test.js
+++ b/test/es-v5/03.basic-querying-no-timeout.test.js
@@ -71,7 +71,7 @@ describe('basic-querying-no-timeout', function () {
         });
 
         PostWithId = db.define('PostWithId', {
-            id: {type: String, id: true}, // TODO: why is this line different from `02.basic-querying.test.js` ?
+            id: {type: String},
             title: {type: String, length: 255},
             content: {type: String}
         });
```

**Comment by pulkitsinghal** [comment]
@aquid - Please go ahead take all the steps required to merge this PR and publish a new release.

Since now we will have `refresh` enabled by default in the esConnector, we should update the **major** version number. If you or anyone else disagrees with me then that's fine, just let me know and carry on.

Definitely should document this fact in release notes regardless:
>  `refresh` is enabled by default in the esConnector for `['create', 'save', 'destroy', 'destroyAll', 'updateAttributes', 'updateOrCreate', 'updateAll']` unless you setup your datasource to prohibit this explicitly with `refreshOn:[]`

There are some improvements left in tests which I've created a new issue for: https://github.com/strongloop-community/loopback-connector-elastic-search/issues/96

That shouldn't stop us from releasing so please go ahead @aquid and make it a great monday for the community!

cc @wolfgang-s @saurabh73 @jamesjara

**Pull request merged**